File: solutions/javascript/bob/1/bob.js
//
// This is only a SKELETON file for the 'Bob' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const hey = (message) => {
  if (isSilence(message)) {
    return 'Fine. Be that way!';
  } else if (isShoutingQuestion(message)) {
    return 'Calm down, I know what I\'m doing!';
  } else if (isShouting(message)) {
    return 'Whoa, chill out!';
  } else if (isQuestion(message)) {
    return 'Sure.';
  } else {
    return 'Whatever.';
  }
};

const isSilence = (message) => {
  return message.trim() === '';
}

const isShouting = (message) => {
  return message.toUpperCase() === message && message.toLowerCase() !== message;
}

const isQuestion = (message) => {
  return message.trim().endsWith('?');
}

const isShoutingQuestion = (message) => {
  return isShouting(message) && isQuestion(message);
}

File: solutions/csharp/bob/1/Bob.cs
using System;
using System.Linq;

public static class Bob
{
    public static string Response(string message) => message switch
    {
        _ when message.IsSilence() => "Fine. Be that way!",
        _ when message.IsShoutingQuestion() => "Calm down, I know what I'm doing!",
        _ when message.IsShouting() => "Whoa, chill out!",
        _ when message.IsQuestion() => "Sure.",
        _ => "Whatever."
    };

    private static bool IsQuestion(this string message)
        => message.TrimEnd().EndsWith("?");

    private static bool IsShouting(this string message)
        => message.Any(char.IsLetter) && message.ToUpperInvariant() == message;

    private static bool IsSilence(this string message)
        => string.IsNullOrWhiteSpace(message);

    private static bool IsShoutingQuestion(this string message)
        => IsQuestion(message) && IsShouting(message);
}